components/speedreader/speedreader_rewriter_service.cc

```diff
@@ -0,0 +1,110 @@
+
+  if (!success || stylesheet.empty()) {
+    VLOG(1) << "Failed to read stylesheet from component: " << stylesheet_path;
+    stylesheet = ui::ResourceBundle::GetSharedInstance().LoadDataResourceString(
+        IDR_SPEEDREADER_STYLE_DESKTOP);
+  }
+
+  return "<style id=\"brave_speedreader_style\">" + stylesheet + "</style>";
+}
+
+}  // namespace
+
+SpeedreaderRewriterService::SpeedreaderRewriterService(
+    brave_component_updater::BraveComponent::Delegate* delegate)
+    : component_(new speedreader::SpeedreaderComponent(delegate)),
+      speedreader_(new speedreader::SpeedReader) {
+  // Load the built-in stylesheet as the default
+  content_stylesheet_ =
+      "<style id=\"brave_speedreader_style\">" +
+      ui::ResourceBundle::GetSharedInstance().LoadDataResourceString(
+          IDR_SPEEDREADER_STYLE_DESKTOP) +
+      "</style>";
+
+  // Check the paths from the component as observer may register
+  // later than the paths were available in the component.
+  const auto stylesheet_path = component_->GetStylesheetPath();
+  if (!stylesheet_path.empty())
+    OnStylesheetReady(stylesheet_path);
+
+  const auto whitelist_path = component_->GetWhitelistPath();
+  if (!whitelist_path.empty())
+    OnWhitelistReady(whitelist_path);
+
+  component_->AddObserver(this);
+}
+
+SpeedreaderRewriterService::~SpeedreaderRewriterService() {
+  component_->RemoveObserver(this);
+}
+
+void SpeedreaderRewriterService::OnWhitelistReady(const base::FilePath& path) {
+  VLOG(2) << "Whitelist ready at " << path;
+  base::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::ThreadPool(), base::MayBlock()},
+      base::BindOnce(
+          &brave_component_updater::LoadDATFileData<speedreader::SpeedReader>,
+          path),
+      base::BindOnce(&SpeedreaderRewriterService::OnLoadDATFileData,
+                     weak_factory_.GetWeakPtr()));
+}
+
+void SpeedreaderRewriterService::OnStylesheetReady(const base::FilePath& path) {
+  base::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::ThreadPool(), base::MayBlock()},
+      base::BindOnce(&GetDistilledPageStylesheet, path),
+      base::BindOnce(&SpeedreaderRewriterService::OnLoadStylesheet,
+                     weak_factory_.GetWeakPtr()));
+}
+
+bool SpeedreaderRewriterService::IsWhitelisted(const GURL& url) {
+  return speedreader_->IsReadableURL(url.spec());
+}
+
+std::unique_ptr<Rewriter> SpeedreaderRewriterService::MakeRewriter(
+    const GURL& url) {
+  return speedreader_->MakeRewriter(url.spec());
+}
+
+const std::string& SpeedreaderRewriterService::GetContentStylesheet() {
+  return content_stylesheet_;
+}
+
+void SpeedreaderRewriterService::OnLoadStylesheet(std::string stylesheet) {
+  VLOG(2) << "Speedreader stylesheet loaded";
+  content_stylesheet_ = stylesheet;
+}
+
+void SpeedreaderRewriterService::OnLoadDATFileData(
+    GetDATFileDataResult result) {
+  VLOG(2) << "Speedreader loaded from DAT file";
+  if (result.first)
+    speedreader_ = std::move(result.first);
```

Comment:
given that in any case we load the whole thing from disk into memory, we could move all work with files / `DAT` stuff to the component and only get final results in this class. But this would make `SpeedreaderComponent` aware of rust "client", so also doesn't look neat

Reply:
it was previously in the component, but this moves it closer to where it is used, and as a next step would allow to more straightforwardly unload both from memory when the feature is disabled independently of component handling